pyproject.toml

```diff
@@ -39,6 +39,9 @@ filterwarnings = [
 ]
 
 [tool.mypy]
+files = "src"
+python_version = "3.6"
```

Comment:
Eh, I don't think that testing against 3.6 only is a particularly good idea and I'd prefer to type check for the Python version I'm developing against by default.

Reply:
I think the best thing is to test for the lowest supported python version. I'd rather always avoid the version currently being used, because you normal tests might already catch that! What I like doing normally (and thought it was done here, but it's not) is to set the minimum in the pyproject.toml, then run the maximum manually also in ~~pre-commit~~ ~~nox~~ tox. That way you are protected from using features that were introduced after 3.6, as well as protected from features that were removed in 3.10 (less common).

Reply:
95% of the time, a feature was added since the (minimum version here) and you doesn't notice it when developing or testing, but mypy can catch it because you are checking against the earliest supported versions.

There are cases were this simply doesn't work, though - like using numpy, which can't type check against 3.6 or even very well against 3.7, and in those cases, you just have to trust your test matrix to catch these sorts of bugs.

Reply:
The version in use is the only version you can reliably type check against because dependencies can (and do) differ. If we want to type check 3.6 we should install Python 3.6 in CI - no gimmicks.

Although, as I recall we limited ourselves to 3.6 because we couldn't get importlib-metadata to behave across all Python versions that we support.

Reply:
This only becomes a problem when something starts dropping the oldest version so the versions mismatch. You can type check quite well against 3.6 from 3.10 as long your dependencies (as installed) still support that version. Typeshed is properly guarded for different versions. This is exactly what I described for NumPy above - it dropped < 3.8, so now you can't typecheck with the latest NumPy and <3.8 as a target, and no NumPy supported <3.7 & types.

Type checking is not runtime, and it's[^1] developer only. If there's a problem, you have to look at it and see if it's real, runtime problem, or a false positive, etc. Practicality beats purity here; you are trying to maximize your ability to find problems. If something comes up (which is rare, _especially with minimal dependencies_, but it can happen), then we can adjust. But for most cases, the Python targeting works extremely well in mypy from other versions.

One difference - I always do type checking in pre-commit, with a (somewhat) locked and very limited environment. This gives more control than assuming some random environment can run type checking. If a dependency doesn't support 3.6, a limit could be included there. And I always specify the minimum here, even if I also test the maximum in CI sometimes.

I'm okay to move this back to CI only, but I think your chances of finding a real error are higher if you have the minimum set than if you don't. Yes, it's not "perfect", but it's rate of failure is lower than the rate of errors caught that would otherwise be missed[^2]. And I am not interested in installing Python 3.6 locally - in fact, on my AS machine, I _can't_, even though I can still type check for 3.6.

[^1]: minus external signatures, but that should not be affected here.
[^2]: I like newer Python features, so maybe my rate of mistakes here is high.

Reply:
Ahh, I see why newer versions were not also run, due to the type ignores. In that case, this really should always be set, because a newer version running the "newer" target will likely tell you to remove type ignores that still need to be there for CI.

Reply:
We can revisit this when we've got static typing to work with all versions that we support. I'm convinced that overriding the default Python version is the better choice for the time being.